drivers/soundwire/stream.c

```diff
@@ -1693,16 +1693,6 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
 		bus->params.bandwidth -= m_rt->stream->params.rate *
 			m_rt->ch_count * m_rt->stream->params.bps;
 
-		/* Compute params */
-		if (bus->compute_params) {
-			ret = bus->compute_params(bus);
-			if (ret < 0) {
-				dev_err(bus->dev, "Compute params failed: %d\n",
-					ret);
-				return ret;
-			}
-		}
-
```

Comment:
The clocks are always updated on prepare/deprepare, I am not sure why you want to remove this.

Reply:
let me re-explain. We have two bank switches, one on prepare and one on trigger. This is intentional, we want to go to the 'new' frequency and let the clock toggle for some time before enabling all the channels on the trigger.

So the point is that we do need to update the parameters and reduce the clock on deprepare. I am not sure why the prepare and de-prepare parts would become asymmetrical?

Reply:
> let me re-explain. We have two bank switches, one on prepare and one on trigger. This is intentional, we want to go to the 'new' frequency and let the clock toggle for some time before enabling all the channels on the trigger.
> 
> So the point is that we do need to update the parameters and reduce the clock on deprepare.

Agree, but m_rt is removed after deprepared and list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) is used by _sdw_compute_port_params(), sdw_compute_group_params(), etc. As a result, the bit slots of the deprepared stream are still reserved. If we reduce the clock, we will get a no sufficient bandwidth error.

> I am not sure why the prepare and de-prepare parts would become asymmetrical?

I don't think it is asymmetrical. Below is the sequence when we start and stop aplay.
intel_hw_params
 -> sdw_stream_add_master
asoc_sdw_prepare
 -> sdw_prepare_stream

asoc_sdw_hw_free
 -> sdw_deprepare_stream
intel_hw_free
 -> sdw_stream_remove_master

Reply:
Not following this. Do we agree when the stream is deprepared, it should use exactly zero bitslots? I don't understand why the bits would remain reserved.

Reply:
> Not following this. Do we agree when the stream is deprepared, it should use exactly zero bitslots? 

Yes, totally agree.

> I don't understand why the bits would remain reserved.

That is the issue in current code and that is because the m_rt is not removed at that point.

Reply:
but the code above still accounts for the bandwidth of the stream, so the stream is still taken into account. 

It makes no sense to me to reduce the bandwidth count *after* performing the bandwidth allocation.

Sorry, completely lost here.

Reply:
Moving the bandwidth reduction count after performing the bandwidth allocation is based on the fact that the m_rt has not been removed when we allocate the bit slots. To fix the issue, I can either move the bandwidth reduction count after performing the bandwidth as the current PR does or remove the m_rt before allocating the bit slots.
I choose the former because removing m_rt first seems break the add m_rt -> prepare -> deprepare ->remove m_rt sequence.
I know the side effects of moving the bandwidth reduction count after performing the bandwidth allocation is that we may keep the sdw bus clock or lane that is not needed. But it will happen only when we reach the threshold that need higher clock rate or more lane. So, I think it is acceptable.

Reply:
That's a bit convoluted really.

The bits for the stream should no longer be accounted for after this routine:

````c
int asoc_sdw_hw_free(struct snd_pcm_substream *substream)
{
	struct snd_soc_pcm_runtime *rtd = snd_soc_substream_to_rtd(substream);
	struct sdw_stream_runtime *sdw_stream;
	struct snd_soc_dai *dai;

	/* Find stream from first CPU DAI */
	dai = snd_soc_rtd_to_cpu(rtd, 0);

	sdw_stream = snd_soc_dai_get_stream(dai, substream->stream);
	if (IS_ERR(sdw_stream)) {
		dev_err(rtd->dev, "no stream found for DAI %s\n", dai->name);
		return PTR_ERR(sdw_stream);
	}

	return sdw_deprepare_stream(sdw_stream);
}
````

but the problem is that we have a half-baked dai handling and comments

````c
static int
intel_hw_free(struct snd_pcm_substream *substream, struct snd_soc_dai *dai)
{
	struct sdw_cdns *cdns = snd_soc_dai_get_drvdata(dai);
	struct sdw_intel *sdw = cdns_to_intel(cdns);
	struct sdw_cdns_dai_runtime *dai_runtime;
	int ret;

	dai_runtime = cdns->dai_runtime_array[dai->id];
	if (!dai_runtime)
		return -EIO;

	/*
	 * The sdw stream state will transition to RELEASED when stream->
	 * master_list is empty. So the stream state will transition to
	 * DEPREPARED for the first cpu-dai and to RELEASED for the last
	 * cpu-dai.
	 */
	ret = sdw_stream_remove_master(&cdns->bus, dai_runtime->stream);
	if (ret < 0) {
		dev_err(dai->dev, "remove master from stream %s failed: %d\n",
			dai_runtime->stream->name, ret);
		return ret;
	}
````

That comment is wrong, the stream state will transition to DEPREPARED in the dailink hw_free(), and then to RELEASED when the last DAI hw_free() is called

That still doesn't really tell me why the bits are not released during the sdw_deprepare_stream() part. There should be no dependency on the m_rt list.

Reply:
asoc_sdw_hw_free() is called before intel_hw_free(). IOW, sdw_deprepare_stream() is called before sdw_stream_remove_master(). So, m_rt is not removed yet when sdw_deprepare_stream() is called.

_sdw_deprepare_stream
-> bus->compute_params(bus)
-> sdw_compute_params()
    -> sdw_compute_bus_params(bus)
    -> sdw_compute_port_params(bus)
````c
                        port_bo = 1;

                        list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
                                sdw_compute_master_ports(m_rt, &params[i], &port_bo, hstop);
                        }

                        hstop = hstop - params[i].hwidth;

````
And we calculate hstart, hstop, and offset in sdw_compute_master_ports()
````c
                t_data.hstart = hstart;
                t_data.hstop = hstop;
                t_data.block_offset = *port_bo;
                t_data.sub_block_offset = 0;
                (*port_bo) += bps * ch;

````
So that we allocate bit slots for each m_rt.

Reply:
yeah, that's clearly wrong, the generic bit allocation should take the states into account, and skip the deprepared streams.

See https://github.com/thesofproject/linux/pull/5177 for compile-tested-only proposal.